### PR TITLE
Enables LDAP-Testing on Travis

### PR DESCRIPTION
**Pull request by heiglandreas**
This PR adds the possibility to run LDAP-tests on travis.

For that a new script is added that sets up an LDAP-server on travis and the phpunit.xml.dist has been adapted accordingly

### CODE REVIEW

**Comment by gianarb** [comment]
Good! :+1:

**Comment by Maks3w** [comment]
@gianarb No it's not good. I don't have time now for to put all the review notes.

File: phpunit.xml.dist [hunk 1]
```diff
@@ -62,22 +62,23 @@
                                 ALT_DN: CN=Alice Baker,OU=Sales,DC=foo,DC=net
                           ALT_PASSWORD: apass
         -->
-        <env name="TESTS_ZEND_LDAP_ONLINE_ENABLED" value="false" />
```
Comment (Maks3w) [inline phpunit.xml.dist]: don't modify this file. Instead add override the env vars in travis.yml

File: test/BindTest.php [hunk 2]
```diff
@@ -185,7 +185,10 @@ public function testRequiresDnWithoutDnBind()
         } catch (Exception\LdapException $zle) {
             /* Note that if your server actually allows anonymous binds this test will fail.
              */
-            $this->assertContains('Failed to retrieve DN', $zle->getMessage());
+            if (! getenv('TESTS_ZEND_LDAP_ANONYMOUS_BIND_ALLOWED')) {
```
Comment (Maks3w) [inline test/BindTest.php]: Put this change in a different PR

File: test/travis/setup_ldap.sh [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+changetype: modify
+replace: olcLogLevel
+olcLogLevel: -1
+#-
+#add: olcDisallows
+#olcDisallows: bind_anon
+EOF
+
+sudo ldapadd -Y EXTERNAL -H ldapi:/// <<EOF
+# database
```
Comment (Maks3w) [inline test/travis/setup_ldap.sh]: Put file contents in his own file

File: test/travis/setup_ldap.sh [hunk 4]
```diff
@@ -0,0 +1,200 @@
+#!/bin/bash
+
+# Script taken from https://github.com/piwik/plugin-LoginLdap/blob/master/tests/travis/setup_ldap.sh
+# install LDAP
+echo "Installing LDAP..."
+sudo apt-get update > /dev/null
+if ! sudo apt-get install slapd ldap-utils -y -qq > /dev/null; then
```
Comment (Maks3w) [inline test/travis/setup_ldap.sh]: Have you tried to do this as described in http://docs.travis-ci.com/user/apt/?

File: .travis.yml [hunk 5]
```diff
@@ -31,11 +31,13 @@ notifications:
   email: false
 
 before_install:
+  - echo "extension=ldap.so" >> `php --ini | grep "Loaded Configuration" | sed -e "s|.*:\s*||"`
```
Comment (Maks3w) [inline .travis.yml]: Use travis config-add instead. See https://github.com/zendframework/zend-cache/blob/master/.travis.yml#L64 for an example

File: phpunit.xml.dist [hunk 6]
```diff
@@ -62,22 +62,23 @@
...
+        <env name="TESTS_ZEND_LDAP_ONLINE_ENABLED" value="true" />
         <env name="TESTS_ZEND_LDAP_HOST" value="localhost" />
         <env name="TESTS_ZEND_LDAP_PORT" value="389" />
-        <env name="TESTS_ZEND_LDAP_USE_START_TLS" value="true" />
+        <env name="TESTS_ZEND_LDAP_USE_START_TLS" value="false" />
         <env name="TESTS_ZEND_LDAP_USE_SSL" value="false" />
-        <env name="TESTS_ZEND_LDAP_USERNAME" value="CN=someUser,DC=example,DC=com" />
-        <env name="TESTS_ZEND_LDAP_PRINCIPAL_NAME" value="someUser@example.com" />
-        <env name="TESTS_ZEND_LDAP_PASSWORD" value="" />
+        <env name="TESTS_ZEND_LDAP_USERNAME" value="cn=deathofrats,dc=ankh-morpork,dc=dw" />
```
Comment (Maks3w) [inline phpunit.xml.dist]: What is the reason for to change the defaults?

**Comment by Maks3w** [comment]
This are the instructions for to setup the LDAP server https://github.com/zendframework/zf2/blob/eb6275e9227e9ed1995d326087488d3ea25811b5/tests/docs/README-LDAP.txt

**Comment by gianarb** [comment]
@Maks3w you are into the review team :D IMO this is a good PR because increase library's value.. :)
You know syntax roles :) Thanks for your review! :+1:

File: test/travis/setup_ldap.sh (hunk 3 above)
Comment (heiglandreas) [inline test/travis/setup_ldap.sh]: I won't be able to use Variable-substitution in those files. Therefore I'd have to change usernames and passwords in there manually which would remove the possibility to adapt to different setups. 

I'd rather avoid that!

File: phpunit.xml.dist [hunk 7]
```diff
@@ -69,7 +69,7 @@
         <env name="TESTS_ZEND_LDAP_USE_SSL" value="false" />
         <env name="TESTS_ZEND_LDAP_USERNAME" value="CN=someUser,DC=example,DC=com" />
         <env name="TESTS_ZEND_LDAP_PRINCIPAL_NAME" value="someUser@example.com" />
-        <env name="TESTS_ZEND_LDAP_PASSWORD" value="" />
+        <env name="TESTS_ZEND_LDAP_PASSWORD" value="insecure" />
```
Comment (heiglandreas) [inline phpunit.xml.dist]: I changed this line as an empty password is not working at all for login. So it's a bad default

File: phpunit.xml.dist [hunk 8]
```diff
@@ -77,7 +77,7 @@
         <env name="TESTS_ZEND_LDAP_ACCOUNT_DOMAIN_NAME_SHORT" value="EXAMPLE" />
         <env name="TESTS_ZEND_LDAP_ALT_USERNAME" value="anotherUser" />
         <env name="TESTS_ZEND_LDAP_ALT_DN" value="CN=Another User,OU=Sales,DC=example,DC=com" />
-        <env name="TESTS_ZEND_LDAP_ALT_PASSWORD" value="" />
+        <env name="TESTS_ZEND_LDAP_ALT_PASSWORD" value="insecure" />
```
Comment (heiglandreas) [inline phpunit.xml.dist]: I changed this line as an empty password is not working at all for login. So it's a bad default

File: README_TEST.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    export TESTS_ZEND_LDAP_PASSWORD="insecure"
+    export TESTS_ZEND_LDAP_BIND_REQUIRES_DN=true
+    export TESTS_ZEND_LDAP_BASE_DN="OU=Sales,DC=example,DC=com"
+    export TESTS_ZEND_LDAP_ACCOUNT_FILTER_FORMAT="(&(objectClass=posixAccount)(uid=%s))"
+    export TESTS_ZEND_LDAP_ACCOUNT_DOMAIN_NAME="example.com"
+    export TESTS_ZEND_LDAP_ACCOUNT_DOMAIN_NAME_SHORT="EXAMPLE"
+    export TESTS_ZEND_LDAP_ALT_USERNAME="anotherUser"
+    export TESTS_ZEND_LDAP_ALT_DN="CN=Another User,OU=Sales,DC=example,DC=com"
+    export TESTS_ZEND_LDAP_ALT_PASSWORD="insecure"
+    export TESTS_ZEND_LDAP_WRITEABLE_SUBTREE="OU=Test,OU=Sales,DC=example,DC=com"
```
Comment (heiglandreas) [inline README_TEST.md]: This is the README_LDAP from https://github.com/zendframework/zf2/blob/eb6275e9227e9ed1995d326087488d3ea25811b5/tests/docs/README-LDAP.txt adapted so that the TestConfiguration Values match those of the phpunit.xml.dist and are set as ENV-variables and not as constants

Comment (Maks3w) [inline README_TEST.md]: Well, use env vars is an option, another is to copy phpunit.xml.dist to phpunit.xml and customize :)

TestConfiguration is the name of the old file TestConfiguration.php.dist used before 2.5 component split.

Comment (heiglandreas) [inline README_TEST.md]: This is just putting the file into the right place (I didn't find it before) and adapting it marginally.

I'll do a complete rewrite in a separate PR!

**Comment by Maks3w** [comment]
This is close. I'll try to fork your PR and adapt it for to not use `sudo` and put ldif files outside of the script.

File: .travis.yml [hunk 10]
```diff
@@ -30,14 +30,31 @@ notifications:
...
   - composer self-update
   - if [[ $EXECUTE_TEST_COVERALLS == 'true' ]]; then composer require --dev --no-update satooshi/php-coveralls ; fi
 
 install:
+  - mkdir -p $TRAVIS_BUILD_DIR/ldap
+  - sudo chmod -R 777 $TRAVIS_BUILD_DIR/ldap
+  - sudo test/travis/configure_ldap.sh
   - travis_retry composer install --no-interaction --ignore-platform-reqs
 
+before_script:
```
Comment (Maks3w) [inline .travis.yml]: Please put this change and .ci/php.ini in a different PR and I'll merge it

**Comment by heiglandreas** [comment]
Am 14.07.15 um 18:50 schrieb Maks3w:
> This is close. I'll try to fork your PR and adapt it for to not use
> |sudo| and put ldif files outside of the script.
> 

I tried to do that (not use |sudo| but I didn't get it to work, which is
somewhat understandable. In the first step of the setup the
LDAP-Admin-user has to be created and that's only possible for the
root-user. As no user exists in LDAP up to that point a privileged user
from outside the LDAP can log into the LDAP-Server and adapt it. That's
where sudo is needed.

But I might be wrong on that and I'd be very glad to learn that it works
without the root-access.

Regarding the ldif-files it's in my eyes far more convenient to be able
to set some of the defaults via environment-variables than to have those
informations hard-coded scattered over different files. Not being able
to use the env-variables setting the password needs to use hardcoded
values where no one is really sure whether they are the actual password
that is documented elsewhere.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zendframework/zend-ldap/pull/8#issuecomment-121305887>.
> 


-- 
                                                              ,,,
                                                             (o o)
+---------------------------------------------------------ooO-(_)-Ooo-+
| Andreas Heigl                                                       |
| mailto:andreas@heigl.org                  N 50°22'59.5" E 08°23'58" |
| http://andreas.heigl.org                       http://hei.gl/wiFKy7 |
+---------------------------------------------------------------------+
| http://hei.gl/root-ca                                               |
+---------------------------------------------------------------------+

**Comment by Maks3w** [comment]
I have found this guy solved the sudo issue by launching a ldap instance in a non reserved port

https://github.com/FreeRADIUS/freeradius-server/pull/1107/files

**Comment by heiglandreas** [comment]
That's great! 

But (you knew there was one coming (-; ) he uses the ```slapd.conf```-version of ldap-configuration instead of the ```slapd-config``` one. The ```slapd.conf```-version is deprecated as http://www.openldap.org/doc/admin24/slapdconfig.html states.  I'm not sure whether the "old" version will be removed any time soon, but...

I can of course use the older ```slapd.conf```-version to configure the server if you think we need to get rid of the sudo-stuff.

**Comment by Maks3w** [comment]
Between use the legacy Travis infrastructure and the legacy LDAP Config I prefer sacrifice slapd.conf and use the deprecated version.

**Comment by heiglandreas** [comment]
Then I'll change that accordingly. But that might take some time as I'm busy until the weekend

**Comment by Maks3w** [comment]
@heiglandreas Don't worry I'm working in my PR adding a Vagrantfile too for easy local testing

**Comment by heiglandreas** [comment]
@Maks3w Ah! Good to know as I was thinking about that as well! You might then use the ldap-script too so we can reduce the amount of files needed!

**Comment by Maks3w** [comment]
My plan is reuse the same commands for Travis and Vagrant

**Comment by heiglandreas** [comment]
:+1:

**Comment by Maks3w** [comment]
Done. Please review #14

**Comment by heiglandreas** [comment]
Doing that tonight!

**Comment by heiglandreas** [comment]
I'll fix this as soon as your PR #14 is merged so I can use the LDAP-scripts in the travis.yml

**Comment by heiglandreas** [comment]
This PR is obsoleted by PR #14
